### PR TITLE
[Serializer][TypeInfo] merge #58804 up to 7.2

### PR DESCRIPTION
**Pull request by mtarld**
| Q             | A
| ------------- | ---
| Branch?       | 7.2
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Issues        | Related to #58796
| License       | MIT

Upmerge #58804 to 7.2

### CODE REVIEW

File: src/Symfony/Component/TypeInfo/TypeResolver/StringTypeResolver.php [hunk 1]
```diff
@@ -58,10 +59,21 @@ final class StringTypeResolver implements TypeResolverInterface
...
     private static array $classExistCache = [];
 
+    private readonly Lexer $lexer;
+    private readonly TypeParser $parser;
+
     public function __construct(
-        private Lexer $lexer = new Lexer(),
-        private TypeParser $parser = new TypeParser(new ConstExprParser()),
+        ?Lexer $lexer = null,
+        ?TypeParser $typeParser = null,
```
Comment (derrabus) [inline src/Symfony/Component/TypeInfo/TypeResolver/StringTypeResolver.php:67]: Those parameters are basically ignored now. Is that what we want?

**Review by derrabus — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/TypeInfo/TypeResolver/StringTypeResolver.php (hunk 1 above)
Comment (mtarld) [reply src/Symfony/Component/TypeInfo/TypeResolver/StringTypeResolver.php:67]: Indeed, my bad

File: src/Symfony/Component/TypeInfo/TypeResolver/PhpDocAwareReflectionTypeResolver.php [hunk 2]
```diff
@@ -33,13 +34,28 @@
...
     public function __construct(
         private TypeResolverInterface $reflectionTypeResolver,
         private TypeResolverInterface $stringTypeResolver,
         private TypeContextFactory $typeContextFactory,
-        private PhpDocParser $phpDocParser = new PhpDocParser(new TypeParser(), new ConstExprParser()),
-        private Lexer $lexer = new Lexer(),
+        ?PhpDocParser $phpDocParser = null,
+        ?Lexer $lexer = null,
     ) {
+        if (!$phpDocParser) {
```
Comment (derrabus) [inline src/Symfony/Component/TypeInfo/TypeResolver/PhpDocAwareReflectionTypeResolver.php:47]: And what if I do pass a `$phpDocParser` to the constructor…?

**Review by derrabus — CHANGES_REQUESTED** [review]

**Comment by derrabus** [comment]
This is my solution now: 72e038eaf36809d01313f9e42675e665baf95b40

Please open a PR if I messed it up. Thank you for your help. ✌🏻

File: src/Symfony/Component/TypeInfo/TypeResolver/PhpDocAwareReflectionTypeResolver.php (hunk 2 above)
Comment (mtarld) [reply src/Symfony/Component/TypeInfo/TypeResolver/PhpDocAwareReflectionTypeResolver.php:47]: Sorry, kind of tired today... Thanks for taking care of that!

**Pull request closed without merging**